src/HappyEyeBallsConnectionBuilder.php

```diff
@@ -0,0 +1,295 @@
+<?php
+
+namespace React\Socket;
+
+use React\Dns\Model\Message;
+use React\Dns\Resolver\Resolver;
+use React\EventLoop\LoopInterface;
+use React\EventLoop\TimerInterface;
+use React\Promise;
+use React\Promise\CancellablePromiseInterface;
+use InvalidArgumentException;
+use RuntimeException;
+
+/**
+ * @internal
+ */
+final class HappyEyeBallsConnectionBuilder
+{
+    public $loop;
+    public $connector;
+    public $resolver;
+    public $uri;
+    public $host;
+    public $resolved = array();
+    public $resolverPromises = array();
+    public $connectionPromises = array();
+    public $connectQueue = array();
+    public $timer;
+    public $parts;
+    public $ipsCount = 0;
+    public $failureCount = 0;
+    public $resolve;
+    public $reject;
+
+    public function __construct(LoopInterface $loop, ConnectorInterface $connector, Resolver $resolver, $uri, $host, $parts)
+    {
+        $this->loop = $loop;
+        $this->connector = $connector;
+        $this->resolver = $resolver;
+        $this->uri = $uri;
+        $this->host = $host;
+        $this->parts = $parts;
+    }
+
+    public function connect()
+    {
+        $that = $this;
+        return new Promise\Promise(function ($resolve, $reject) use ($that) {
+            /**
+             * Delay A lookup by 50ms as per RFC.
+             *
+             * @link https://tools.ietf.org/html/rfc8305#section-3
+             */
+            $that->resolve(Message::TYPE_AAAA, $resolve, $reject);
+            Promise\Timer\resolve(0.05, $that->loop)->then(function () use ($that, $resolve, $reject) {
+                $that->resolve(Message::TYPE_A, $resolve, $reject);
+            });
```

Comment:
It looks like this piece of logic is missing some cancellation handling, it should probably stop the timer or the lookup? What happens if I cancel the resulting promise? What happens if the IPv6 connection has succeeded already, it should probably stop the timer and not send the second lookup?

Reply:
It will still send the second look up. Just added cancelation logic